tap_appstore/client.py

```diff
@@ -18,13 +23,140 @@
+
+    def update_stream_state(self, date):
+        """Update the stream state with the new date."""
+        self.stream_state['start_date'] = date.strftime('%Y-%m-%d')
+        self.logger.info(f"Updating state, new start date is {self.stream_state['start_date']}")
+
+    def get_records(self, context: dict = None):
+        start_date = self.get_start_date(default_date='2024-01-01')
+        if not start_date:
+            logger.error("Start date could not be determined.")
+            return
+
+        start_date = start_date.replace(tzinfo=None)
+        date_limit = datetime.now().replace(tzinfo=None) - timedelta(days=2)
+        line_id = 0
+
+        while start_date <= date_limit:
+
+            try:
+                logger.info(f"start_date_date_only: {start_date.strftime('%Y-%m-%d')}")
+                all_data = self.download_data(start_date, self.api)
+
+                data_io = StringIO(all_data)
+
+                # Rename column names from 'App Name' -> 'app_name'
+                first_line = data_io.readline().strip()
+                fieldnames = [col.strip().replace(' ', '_').lower() for col in first_line.split('\t')]
+
+                # Load TSV from memory
+                reader = csv.DictReader(data_io, delimiter='\t', fieldnames=fieldnames)
+
+                for record in reader:
+                    line_id += 1
+                    record['_line_id'] = line_id
+                    record['_time_extracted'] = datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ')
+                    record['_api_report_date'] = self.get_report_date(start_date)
+
+                    processed_record = self.process_record(record)
+                    if processed_record is not None:
+                        yield processed_record
+
+                    yield record
+
+            except APIError as e:
+                logger.error(f'Error during download report {self.name}.\n{e}')
+
+            start_date = self.increment_date(start_date)
+            self.update_stream_state(start_date)
+
+
+    @staticmethod
+    def convert_date(date_str, date_format='%Y-%m-%d'):
+        """Converts date string to ISO format based on the given date format, defaulting to '%Y-%m-%d'.
+           Returns None if the date string is invalid or empty.
+        """
+
+        if not date_str:
+            logger.warning(f"Empty or invalid date string provided; cannot convert using format {date_format}")
+            return None
+
+        date_str = date_str.strip()
+        if not date_str:
+            logger.warning(f"Empty or invalid date string provided; cannot convert using format {date_format}")
+            return None
+
+        try:
+            return datetime.strptime(date_str, date_format).isoformat()
+        except ValueError as e:
+            logger.error(f"Date conversion error for date: '{date_str}' with format '{date_format}' | Error: {str(e)}")
+            return None
+
+    def convert_fields(self, record, fields, target_type):
```

Comment:
I have included this for now, as if we don't explicitly call `int()` or `float(()` on `IntegerType` and `NumberType` fields, I am seeing errors like this from Meltano:

`jsonschema.exceptions.ValidationError: '0.00' is not of type 'number', 'null'`

`jsonschema.exceptions.ValidationError: '1075' is not of type 'integer', 'null'`

Not sure if there is a better way to do this.

Reply:
This is the validation step, since we are reading from a TSV, it will always read as string. Let's force the conversion in the post-processing.

Reply:
The validation is important when we are reading the records directly from an API.

Reply:
> Let's force the conversion in the post-processing instead of doing it in the base Stream.

We do this in  the `post_process()` function now!